torch/_inductor/optimize_indexing.py

```diff
@@ -116,3 +116,17 @@ def indexing_dtype_strength_reduction(loop_body: LoopBody):
             loop_body.indexing_exprs,
             bv.replacement_vals,
         )
+
+
+def all_indexing_expressable_in_32_bits(loop_body: LoopBody) -> bool:
+    """
+    Are all get_index and index_expr nodes values expressable in 32 bits
+    """
+    bv = loop_body.bounds().get_bounds()
+    for node in loop_body.get_nodes():
+        if node.target in ("get_index", "index_expr"):
+            if vr := bv.get(node, None):
+                if not range_expressable_in_32_bits(vr):
```

Comment:
The current strategy is to use the hint values to determine if we should generate a 32-bit indexing kernel, and generate guards appropriately. This means we get optimized kernels even if the range analysis can't prove an upper bound, which happens with basically any dynamic shape.

I'm not sure relying on range analysis is going to work at all tbh.

Reply:
Yea, didn't forgot range analysis bails on dynamic shapes.